TestCascadingDeletion flaky [timed out waiting for the condition]

<!-- This form is for bug reports and feature requests ONLY! 

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:


/kind bug



**What happened**:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/56097/pull-kubernetes-unit/67068/
```
go test -v k8s.io/kubernetes/test/integration/garbagecollector -run TestCascadingDeletion$
W1120 23:43:24.574825   19542 authentication.go:56] Authentication is disabled
I1120 23:43:24.581939   19542 serve.go:86] Serving securely on 127.0.0.1:39739
I1120 23:43:24.582037   19542 crd_finalizer.go:242] Starting CRDFinalizer
I1120 23:43:24.582077   19542 customresource_discovery_controller.go:152] Starting DiscoveryController
I1120 23:43:24.582117   19542 naming_controller.go:274] Starting NamingConditionController
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"apiextensions.k8s.io/v1beta1","resources":[{"name":"customresourcedefinitions","singularName":"","namespaced":false,"kind":"CustomResourceDefinition","verbs":["create","delete","deletecollection","get","list","patch","update","watch"],"shortNames":["crd"]},{"name":"customresourcedefinitions/status","singularName":"","namespaced":false,"kind":"CustomResourceDefinition","verbs":["update"]}]}
I1120 23:43:24.614636   19542 crd_finalizer.go:246] Shutting down CRDFinalizer
I1120 23:43:24.614731   19542 naming_controller.go:278] Shutting down NamingConditionController
E1120 23:43:24.614641   19542 customresource_discovery_controller.go:155] timed out waiting for caches to sync
I1120 23:43:24.615316   19542 customresource_discovery_controller.go:156] Shutting down DiscoveryController
I1120 23:43:24.614679   19542 serve.go:137] Stopped listening on 127.0.0.1:39739
```
**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:


**Anything else we need to know?**:

`E1120 23:43:24.614641   19542 customresource_discovery_controller.go:155] timed out waiting for caches to sync` This should not occur unless stopCh closed. But no more log. And from the ci log, we can see the kube-apiserever is up and down for many times, So I do not know why. 

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/56126/pull-kubernetes-unit/68031/artifacts/junit_e9d37341e9f7ab19ac436e5365ba6ae7c20eea71_20171124-060937.stdout

It seems rc not garbage collected after calling delete.  I do not know why now, can any one help. 

```
I1124 06:13:02.907755   19653 garbagecollector.go:408] object [v1/ReplicationController, namespace: gc-cascading-deletion, name: test.rc.1, uid: 8715d1e5-d0de-11e7-8b29-0242ac110002]'s doesn't have an owner, continue on next item
...
--- FAIL: TestCascadingDeletion (67.05s)
	testserver.go:100: Starting kube-apiserver on port 39557...
	testserver.go:112: Waiting for /healthz to be ok...
	garbage_collector_test.go:366: timed out waiting for the condition
```

still seeing TestCascadingDeletion failures after #56466 merged

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/56538/pull-kubernetes-unit/68725/

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/56513/pull-kubernetes-unit/68675/

```
--- FAIL: TestCascadingDeletion (69.08s)
	testserver.go:100: Starting kube-apiserver on port 40501...
	testserver.go:112: Waiting for /healthz to be ok...
	garbage_collector_test.go:366: timed out waiting for the condition
```

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/57412/pull-kubernetes-unit/71699/artifacts/junit_e9d37341e9f7ab19ac436e5365ba6ae7c20eea71_20171219-164009.stdout
<details>
```
=== RUN   TestCascadingDeletion
I1219 16:44:40.098299   23324 apiservice_controller.go:124] Shutting down APIServiceRegistrationController
I1219 16:44:40.098360   23324 available_controller.go:274] Shutting down AvailableConditionController
I1219 16:44:40.098402   23324 customresource_discovery_controller.go:163] Shutting down DiscoveryController
I1219 16:44:40.098427   23324 naming_controller.go:285] Shutting down NamingConditionController
I1219 16:44:40.098447   23324 crdregistration_controller.go:139] Shutting down crd-autoregister controller
I1219 16:44:40.098427   23324 autoregister_controller.go:160] Shutting down autoregister controller
I1219 16:44:40.098459   23324 crd_finalizer.go:254] Shutting down CRDFinalizer
I1219 16:44:40.098542   23324 controller.go:90] Shutting down OpenAPI AggregationController
I1219 16:44:40.098751   23324 serve.go:129] Stopped listening on [::]:34363
I1219 16:44:40.453915   23324 serving.go:295] Generated self-signed cert (/tmp/kubernetes-kube-apiserver709151440/apiserver.crt, /tmp/kubernetes-kube-apiserver709151440/apiserver.key)
W1219 16:44:40.453952   23324 authentication.go:378] AnonymousAuth is not allowed with the AllowAll authorizer.  Resetting AnonymousAuth to false. You should use a different authorizer
I1219 16:44:40.605180   23324 master.go:225] Using reconciler: master-count
[restful] 2017/12/19 16:44:40 log.go:33: [restful/swagger] listing is available at https://172.17.0.2:45589/swaggerapi
[restful] 2017/12/19 16:44:40 log.go:33: [restful/swagger] https://172.17.0.2:45589/swaggerui/ is mapped to folder /swagger-ui/
[restful] 2017/12/19 16:44:42 log.go:33: [restful/swagger] listing is available at https://172.17.0.2:45589/swaggerapi
[restful] 2017/12/19 16:44:42 log.go:33: [restful/swagger] https://172.17.0.2:45589/swaggerui/ is mapped to folder /swagger-ui/
I1219 16:44:47.340818   23324 serve.go:89] Serving securely on [::]:45589
I1219 16:44:47.340885   23324 apiservice_controller.go:112] Starting APIServiceRegistrationController
I1219 16:44:47.340898   23324 cache.go:32] Waiting for caches to sync for APIServiceRegistrationController controller
I1219 16:44:47.341741   23324 crd_finalizer.go:242] Starting CRDFinalizer
I1219 16:44:47.342442   23324 customresource_discovery_controller.go:152] Starting DiscoveryController
I1219 16:44:47.342479   23324 naming_controller.go:274] Starting NamingConditionController
I1219 16:44:47.342618   23324 controller.go:84] Starting OpenAPI AggregationController
I1219 16:44:47.343019   23324 available_controller.go:262] Starting AvailableConditionController
I1219 16:44:47.343044   23324 cache.go:32] Waiting for caches to sync for AvailableConditionController controller
I1219 16:44:47.343495   23324 crdregistration_controller.go:110] Starting crd-autoregister controller
I1219 16:44:47.343525   23324 controller_utils.go:1019] Waiting for caches to sync for crd-autoregister controller
I1219 16:44:47.441127   23324 cache.go:39] Caches are synced for APIServiceRegistrationController controller
I1219 16:44:47.443223   23324 cache.go:39] Caches are synced for AvailableConditionController controller
I1219 16:44:47.443660   23324 controller_utils.go:1026] Caches are synced for crd-autoregister controller
I1219 16:44:47.443769   23324 autoregister_controller.go:136] Starting autoregister controller
I1219 16:44:47.443824   23324 cache.go:32] Waiting for caches to sync for autoregister controller
I1219 16:44:47.544149   23324 cache.go:39] Caches are synced for autoregister controller
I1219 16:44:47.738151   23324 graph_builder.go:190] using a shared informer for resource "apps/v1, Resource=daemonsets", kind "apps/v1, Kind=DaemonSet"
I1219 16:44:47.738796   23324 graph_builder.go:190] using a shared informer for resource "rbac.authorization.k8s.io/v1, Resource=rolebindings", kind "rbac.authorization.k8s.io/v1, Kind=RoleBinding"
I1219 16:44:47.738929   23324 graph_builder.go:190] using a shared informer for resource "scheduling.k8s.io/v1alpha1, Resource=priorityclasses", kind "scheduling.k8s.io/v1alpha1, Kind=PriorityClass"
I1219 16:44:47.739069   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=services", kind "/v1, Kind=Service"
I1219 16:44:47.739155   23324 graph_builder.go:190] using a shared informer for resource "apps/v1, Resource=deployments", kind "apps/v1, Kind=Deployment"
I1219 16:44:47.739294   23324 graph_builder.go:190] using a shared informer for resource "storage.k8s.io/v1, Resource=storageclasses", kind "storage.k8s.io/v1, Kind=StorageClass"
I1219 16:44:47.739419   23324 graph_builder.go:190] using a shared informer for resource "admissionregistration.k8s.io/v1beta1, Resource=mutatingwebhookconfigurations", kind "admissionregistration.k8s.io/v1beta1, Kind=MutatingWebhookConfiguration"
I1219 16:44:47.739616   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=replicationcontrollers", kind "/v1, Kind=ReplicationController"
I1219 16:44:47.739731   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=endpoints", kind "/v1, Kind=Endpoints"
I1219 16:44:47.739852   23324 graph_builder.go:190] using a shared informer for resource "batch/v1beta1, Resource=cronjobs", kind "batch/v1beta1, Kind=CronJob"
I1219 16:44:47.739962   23324 graph_builder.go:190] using a shared informer for resource "policy/v1beta1, Resource=poddisruptionbudgets", kind "policy/v1beta1, Kind=PodDisruptionBudget"
I1219 16:44:47.740078   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=pods", kind "/v1, Kind=Pod"
I1219 16:44:47.740216   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=resourcequotas", kind "/v1, Kind=ResourceQuota"
I1219 16:44:47.740405   23324 graph_builder.go:190] using a shared informer for resource "storage.k8s.io/v1alpha1, Resource=volumeattachments", kind "storage.k8s.io/v1alpha1, Kind=VolumeAttachment"
I1219 16:44:47.740630   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=persistentvolumes", kind "/v1, Kind=PersistentVolume"
I1219 16:44:47.740757   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=serviceaccounts", kind "/v1, Kind=ServiceAccount"
I1219 16:44:47.740894   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=nodes", kind "/v1, Kind=Node"
I1219 16:44:47.741109   23324 graph_builder.go:190] using a shared informer for resource "extensions/v1beta1, Resource=ingresses", kind "extensions/v1beta1, Kind=Ingress"
I1219 16:44:47.741210   23324 graph_builder.go:190] using a shared informer for resource "autoscaling/v1, Resource=horizontalpodautoscalers", kind "autoscaling/v1, Kind=HorizontalPodAutoscaler"
I1219 16:44:47.741338   23324 graph_builder.go:190] using a shared informer for resource "settings.k8s.io/v1alpha1, Resource=podpresets", kind "settings.k8s.io/v1alpha1, Kind=PodPreset"
I1219 16:44:47.741439   23324 graph_builder.go:190] using a shared informer for resource "apps/v1, Resource=controllerrevisions", kind "apps/v1, Kind=ControllerRevision"
I1219 16:44:47.741580   23324 graph_builder.go:190] using a shared informer for resource "events.k8s.io/v1beta1, Resource=events", kind "events.k8s.io/v1beta1, Kind=Event"
I1219 16:44:47.741703   23324 graph_builder.go:190] using a shared informer for resource "admissionregistration.k8s.io/v1beta1, Resource=validatingwebhookconfigurations", kind "admissionregistration.k8s.io/v1beta1, Kind=ValidatingWebhookConfiguration"
I1219 16:44:47.741960   23324 graph_builder.go:190] using a shared informer for resource "rbac.authorization.k8s.io/v1, Resource=clusterrolebindings", kind "rbac.authorization.k8s.io/v1, Kind=ClusterRoleBinding"
I1219 16:44:47.742128   23324 graph_builder.go:190] using a shared informer for resource "rbac.authorization.k8s.io/v1, Resource=roles", kind "rbac.authorization.k8s.io/v1, Kind=Role"
I1219 16:44:47.742245   23324 graph_builder.go:190] using a shared informer for resource "admissionregistration.k8s.io/v1alpha1, Resource=initializerconfigurations", kind "admissionregistration.k8s.io/v1alpha1, Kind=InitializerConfiguration"
I1219 16:44:47.742462   23324 graph_builder.go:195] unable to use a shared informer for resource "extensions/v1beta1, Resource=networkpolicies", kind "extensions/v1beta1, Kind=NetworkPolicy": no informer found for {extensions v1beta1 networkpolicies}
I1219 16:44:47.742507   23324 graph_builder.go:199] create storage for resource {extensions v1beta1 networkpolicies}
I1219 16:44:47.742789   23324 graph_builder.go:190] using a shared informer for resource "extensions/v1beta1, Resource=deployments", kind "extensions/v1beta1, Kind=Deployment"
I1219 16:44:47.742921   23324 graph_builder.go:190] using a shared informer for resource "batch/v1, Resource=jobs", kind "batch/v1, Kind=Job"
I1219 16:44:47.743083   23324 graph_builder.go:190] using a shared informer for resource "rbac.authorization.k8s.io/v1, Resource=clusterroles", kind "rbac.authorization.k8s.io/v1, Kind=ClusterRole"
I1219 16:44:47.743353   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=configmaps", kind "/v1, Kind=ConfigMap"
I1219 16:44:47.743575   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=namespaces", kind "/v1, Kind=Namespace"
I1219 16:44:47.743804   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=podtemplates", kind "/v1, Kind=PodTemplate"
I1219 16:44:47.743988   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=limitranges", kind "/v1, Kind=LimitRange"
I1219 16:44:47.744220   23324 graph_builder.go:190] using a shared informer for resource "networking.k8s.io/v1, Resource=networkpolicies", kind "networking.k8s.io/v1, Kind=NetworkPolicy"
I1219 16:44:47.744437   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=secrets", kind "/v1, Kind=Secret"
I1219 16:44:47.744642   23324 graph_builder.go:190] using a shared informer for resource "/v1, Resource=persistentvolumeclaims", kind "/v1, Kind=PersistentVolumeClaim"
I1219 16:44:47.744803   23324 graph_builder.go:190] using a shared informer for resource "extensions/v1beta1, Resource=replicasets", kind "extensions/v1beta1, Kind=ReplicaSet"
I1219 16:44:47.744917   23324 graph_builder.go:190] using a shared informer for resource "extensions/v1beta1, Resource=podsecuritypolicies", kind "extensions/v1beta1, Kind=PodSecurityPolicy"
I1219 16:44:47.745111   23324 graph_builder.go:190] using a shared informer for resource "extensions/v1beta1, Resource=daemonsets", kind "extensions/v1beta1, Kind=DaemonSet"
I1219 16:44:47.745332   23324 graph_builder.go:190] using a shared informer for resource "apps/v1, Resource=replicasets", kind "apps/v1, Kind=ReplicaSet"
I1219 16:44:47.745498   23324 graph_builder.go:190] using a shared informer for resource "apps/v1, Resource=statefulsets", kind "apps/v1, Kind=StatefulSet"
I1219 16:44:47.745698   23324 graph_builder.go:190] using a shared informer for resource "certificates.k8s.io/v1beta1, Resource=certificatesigningrequests", kind "certificates.k8s.io/v1beta1, Kind=CertificateSigningRequest"
I1219 16:44:47.745796   23324 graph_builder.go:263] synced monitors; added 43, kept 0, removed 0
I1219 16:44:47.745999   23324 garbagecollector.go:136] Starting garbage collector controller
I1219 16:44:47.748325   23324 controller_utils.go:1019] Waiting for caches to sync for garbage collector controller
I1219 16:44:47.748357   23324 graph_builder.go:321] GraphBuilder running
I1219 16:44:47.748702   23324 graph_builder.go:295] started 43 new monitors, 43 currently running
I1219 16:44:47.774874   23324 graph_builder.go:588] GraphBuilder process object: v1/Namespace, namespace , name default, uid ec6dd62e-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:47.774992   23324 graph_builder.go:588] GraphBuilder process object: v1/Namespace, namespace , name kube-system, uid ec70574d-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:47.775397   23324 graph_builder.go:588] GraphBuilder process object: v1/Namespace, namespace , name kube-public, uid ec70e4ec-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:47.775494   23324 graph_builder.go:588] GraphBuilder process object: v1/Namespace, namespace , name aval, uid ec99f4da-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:47.776594   23324 graph_builder.go:588] GraphBuilder process object: v1/Endpoints, namespace default, name kubernetes, uid ec6f50cf-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:47.786604   23324 graph_builder.go:588] GraphBuilder process object: v1/Namespace, namespace , name gc-cascading-deletion, uid ecace303-e4db-11e7-88ab-0242ac110002, event type add
W1219 16:44:47.799199   23324 reflector.go:341] k8s.io/kubernetes/vendor/k8s.io/client-go/informers/factory.go:86: watch of *v1.ReplicationController ended with: too old resource version: 4291 (4860)
I1219 16:44:47.880786   23324 graph_builder.go:588] GraphBuilder process object: v1/Service, namespace default, name kubernetes, uid ec6ead45-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:48.017606   23324 graph_builder.go:588] GraphBuilder process object: v1/ServiceAccount, namespace aval, name default, uid ec9aa443-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:48.017668   23324 graph_builder.go:588] GraphBuilder process object: v1/ServiceAccount, namespace gc-cascading-deletion, name default, uid ecae361b-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:48.148587   23324 controller_utils.go:1026] Caches are synced for garbage collector controller
I1219 16:44:48.148629   23324 garbagecollector.go:145] Garbage collector: all resource monitors have synced. Proceeding to collect garbage
I1219 16:44:48.199224   23324 graph_builder.go:588] GraphBuilder process object: v1/Pod, namespace gc-cascading-deletion, name test.pod.1, uid ecec42db-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:48.199297   23324 graph_builder.go:396] add virtual node.identity: [v1/ReplicationController, namespace: gc-cascading-deletion, name: test.rc.1, uid: ecaed94f-e4db-11e7-88ab-0242ac110002]

I1219 16:44:48.199391   23324 garbagecollector.go:374] processing item [v1/ReplicationController, namespace: gc-cascading-deletion, name: test.rc.1, uid: ecaed94f-e4db-11e7-88ab-0242ac110002]
I1219 16:44:48.209573   23324 garbagecollector.go:411] object [v1/ReplicationController, namespace: gc-cascading-deletion, name: test.rc.1, uid: ecaed94f-e4db-11e7-88ab-0242ac110002]'s doesn't have an owner, continue on next item
I1219 16:44:48.238847   23324 graph_builder.go:588] GraphBuilder process object: v1/Pod, namespace gc-cascading-deletion, name test.pod.3, uid ecf26de2-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:48.238913   23324 graph_builder.go:396] add virtual node.identity: [v1/ReplicationController, namespace: gc-cascading-deletion, name: test.rc.2, uid: ecaf8bc0-e4db-11e7-88ab-0242ac110002]

I1219 16:44:48.238970   23324 garbagecollector.go:374] processing item [v1/ReplicationController, namespace: gc-cascading-deletion, name: test.rc.2, uid: ecaf8bc0-e4db-11e7-88ab-0242ac110002]
I1219 16:44:48.241206   23324 garbagecollector.go:411] object [v1/ReplicationController, namespace: gc-cascading-deletion, name: test.rc.2, uid: ecaf8bc0-e4db-11e7-88ab-0242ac110002]'s doesn't have an owner, continue on next item
I1219 16:44:48.279013   23324 graph_builder.go:588] GraphBuilder process object: v1/Pod, namespace gc-cascading-deletion, name test.pod.2, uid ecf876df-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:48.657159   23324 garbagecollector.go:183] syncing garbage collector with updated resources from discovery: map[{ v1 events}:{} {rbac.authorization.k8s.io v1 roles}:{} {rbac.authorization.k8s.io v1 clusterrolebindings}:{} {storage.k8s.io v1alpha1 volumeattachments}:{} { v1 replicationcontrollers}:{} { v1 serviceaccounts}:{} { v1 limitranges}:{} {extensions v1beta1 ingresses}:{} {extensions v1beta1 daemonsets}:{} {apps v1 controllerrevisions}:{} {policy v1beta1 poddisruptionbudgets}:{} {settings.k8s.io v1alpha1 podpresets}:{} { v1 resourcequotas}:{} {apps v1 daemonsets}:{} {apps v1 replicasets}:{} {certificates.k8s.io v1beta1 certificatesigningrequests}:{} {admissionregistration.k8s.io v1beta1 validatingwebhookconfigurations}:{} { v1 configmaps}:{} {apps v1 statefulsets}:{} {admissionregistration.k8s.io v1beta1 mutatingwebhookconfigurations}:{} {scheduling.k8s.io v1alpha1 priorityclasses}:{} { v1 pods}:{} {extensions v1beta1 networkpolicies}:{} {batch v1beta1 cronjobs}:{} {networking.k8s.io v1 networkpolicies}:{} { v1 persistentvolumes}:{} { v1 secrets}:{} { v1 namespaces}:{} {extensions v1beta1 replicasets}:{} {events.k8s.io v1beta1 events}:{} {batch v1 jobs}:{} {rbac.authorization.k8s.io v1 rolebindings}:{} {admissionregistration.k8s.io v1alpha1 initializerconfigurations}:{} {extensions v1beta1 deployments}:{} {apps v1 deployments}:{} {storage.k8s.io v1 storageclasses}:{} { v1 podtemplates}:{} { v1 nodes}:{} { v1 persistentvolumeclaims}:{} { v1 services}:{} {apiregistration.k8s.io v1beta1 apiservices}:{} {extensions v1beta1 podsecuritypolicies}:{} {autoscaling v1 horizontalpodautoscalers}:{} {rbac.authorization.k8s.io v1 clusterroles}:{} {apiextensions.k8s.io v1beta1 customresourcedefinitions}:{} { v1 endpoints}:{}]
I1219 16:44:48.837813   23324 graph_builder.go:588] GraphBuilder process object: v1/ReplicationController, namespace gc-cascading-deletion, name test.rc.2, uid ecaf8bc0-e4db-11e7-88ab-0242ac110002, event type add
I1219 16:44:49.316790   23324 graph_builder.go:263] synced monitors; added 0, kept 43, removed 0
I1219 16:44:49.316840   23324 graph_builder.go:295] started 0 new monitors, 43 currently running
I1219 16:44:49.316860   23324 controller_utils.go:1019] Waiting for caches to sync for garbage collector controller
I1219 16:44:49.417177   23324 controller_utils.go:1026] Caches are synced for garbage collector controller
I1219 16:44:49.417222   23324 garbagecollector.go:220] synced garbage collector
I1219 16:44:54.447225   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:44:59.473854   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:04.500785   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:09.530108   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:14.562200   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:19.590412   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:24.611082   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:29.640088   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:34.663912   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:39.691322   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:44.717282   23324 garbagecollector.go:178] no resource updates from discovery, skipping garbage collector sync
I1219 16:45:48.373161   23324 garbagecollector.go:154] Shutting down garbage collector controller
I1219 16:45:48.377971   23324 graph_builder.go:348] stopped 43 of 43 monitors
I1219 16:45:48.378008   23324 graph_builder.go:349] GraphBuilder stopping
--- FAIL: TestCascadingDeletion (68.31s)
	testserver.go:100: Starting kube-apiserver on port 45589...
	testserver.go:112: Waiting for /healthz to be ok...
	garbage_collector_test.go:368: timed out waiting for the condition
```
</details>

--- a/pkg/controller/garbagecollector/garbagecollector.go
+++ b/pkg/controller/garbagecollector/garbagecollector.go
@@ -38,7 +38,6 @@ import (
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/kubernetes/pkg/controller"
-	"k8s.io/kubernetes/pkg/controller/garbagecollector/metaonly"
 	_ "k8s.io/kubernetes/pkg/util/reflector/prometheus" // for reflector metric registration
 	// install the prometheus plugin
 	_ "k8s.io/kubernetes/pkg/util/workqueue/prometheus"
@@ -259,22 +258,14 @@ func (gc *GarbageCollector) attemptToDeleteWorker() bool {
 		}
 		// retry if garbage collection of an object failed.
 		gc.attemptToDelete.AddRateLimited(item)
+	} else if !n.isObserved() {
+		// requeue if item hasn't been observed via an informer event yet.
+		// otherwise a virtual node for an item added AND removed during watch reestablishment can get stuck in the graph and never removed.
+		// see https://issue.k8s.io/56121
+		glog.V(5).Infof("item %s hasn't been observed via informer yet", n.identity)
+		gc.attemptToDelete.AddRateLimited(item)
 	}
 	return true
-}
-
-func objectReferenceToMetadataOnlyObject(ref objectReference) *metaonly.MetadataOnlyObject {
-	return &metaonly.MetadataOnlyObject{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: ref.APIVersion,
-			Kind:       ref.Kind,
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Namespace: ref.Namespace,
-			UID:       ref.UID,
-			Name:      ref.Name,
-		},
-	}
 }
 
 // isDangling check if a reference is pointing to an object that doesn't exist.
@@ -353,15 +344,6 @@ func (gc *GarbageCollector) classifyReferences(item *node, latestReferences []me
 	return solid, dangling, waitingForDependentsDeletion, nil
 }
 
-func (gc *GarbageCollector) generateVirtualDeleteEvent(identity objectReference) {
-	event := &event{
-		eventType: deleteEvent,
-		obj:       objectReferenceToMetadataOnlyObject(identity),
-	}
-	glog.V(5).Infof("generating virtual delete event for %s\n\n", event.obj)
-	gc.dependencyGraphBuilder.enqueueChanges(event)
-}
-
 func ownerRefsToUIDs(refs []metav1.OwnerReference) []types.UID {
 	var ret []types.UID
 	for _, ref := range refs {
@@ -387,7 +369,10 @@ func (gc *GarbageCollector) attemptToDeleteItem(item *node) error {
 		// exist yet, so we need to enqueue a virtual Delete event to remove
 		// the virtual node from GraphBuilder.uidToNode.
 		glog.V(5).Infof("item %v not found, generating a virtual delete event", item.identity)
-		gc.generateVirtualDeleteEvent(item.identity)
+		gc.dependencyGraphBuilder.enqueueVirtualDeleteEvent(item.identity)
+		// since we're manually inserting a delete event to remove this node,
+		// we don't need to keep tracking it as a virtual node and requeueing in attemptToDelete
+		item.markObserved()
 		return nil
 	case err != nil:
 		return err
@@ -395,7 +380,10 @@ func (gc *GarbageCollector) attemptToDeleteItem(item *node) error {
 
 	if latest.GetUID() != item.identity.UID {
 		glog.V(5).Infof("UID doesn't match, item %v not found, generating a virtual delete event", item.identity)
-		gc.generateVirtualDeleteEvent(item.identity)
+		gc.dependencyGraphBuilder.enqueueVirtualDeleteEvent(item.identity)
+		// since we're manually inserting a delete event to remove this node,
+		// we don't need to keep tracking it as a virtual node and requeueing in attemptToDelete
+		item.markObserved()
 		return nil
 	}
 

--- a/pkg/controller/garbagecollector/graph.go
+++ b/pkg/controller/garbagecollector/graph.go
@@ -53,6 +53,9 @@ type node struct {
 	// this records if the object's deletionTimestamp is non-nil.
 	beingDeleted     bool
 	beingDeletedLock sync.RWMutex
+	// this records if the object was constructed virtually and never observed via informer event
+	virtual     bool
+	virtualLock sync.RWMutex
 	// when processing an Update event, we need to compare the updated
 	// ownerReferences with the owners recorded in the graph.
 	owners []metav1.OwnerReference
@@ -70,6 +73,17 @@ func (n *node) isBeingDeleted() bool {
 	n.beingDeletedLock.RLock()
 	defer n.beingDeletedLock.RUnlock()
 	return n.beingDeleted
+}
+
+func (n *node) markObserved() {
+	n.virtualLock.Lock()
+	defer n.virtualLock.Unlock()
+	n.virtual = false
+}
+func (n *node) isObserved() bool {
+	n.virtualLock.RLock()
+	defer n.virtualLock.RUnlock()
+	return n.virtual == false
 }
 
 func (n *node) markDeletingDependents() {

--- a/pkg/controller/garbagecollector/graph_builder.go
+++ b/pkg/controller/garbagecollector/graph_builder.go
@@ -37,6 +37,7 @@ import (
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/controller/garbagecollector/metaonly"
 )
 
 type eventType int
@@ -369,8 +370,16 @@ func DefaultIgnoredResources() map[schema.GroupResource]struct{} {
 	return ignoredResources
 }
 
-func (gb *GraphBuilder) enqueueChanges(e *event) {
-	gb.graphChanges.Add(e)
+// enqueueVirtualDeleteEvent is used to add a virtual delete event to be processed for virtual nodes
+// once it is determined they do not have backing objects in storage
+func (gb *GraphBuilder) enqueueVirtualDeleteEvent(ref objectReference) {
+	gb.graphChanges.Add(&event{
+		eventType: deleteEvent,
+		obj: &metaonly.MetadataOnlyObject{
+			TypeMeta:   metav1.TypeMeta{APIVersion: ref.APIVersion, Kind: ref.Kind},
+			ObjectMeta: metav1.ObjectMeta{Namespace: ref.Namespace, UID: ref.UID, Name: ref.Name},
+		},
+	})
 }
 
 // addDependentToOwners adds n to owners' dependents list. If the owner does not
@@ -389,6 +398,7 @@ func (gb *GraphBuilder) addDependentToOwners(n *node, owners []metav1.OwnerRefer
 					Namespace:      n.identity.Namespace,
 				},
 				dependents: make(map[*node]struct{}),
+				virtual:    true,
 			}
 			glog.V(5).Infof("add virtual node.identity: %s\n\n", ownerNode.identity)
 			gb.uidToNode.Write(ownerNode)
@@ -591,6 +601,12 @@ func (gb *GraphBuilder) processGraphChanges() bool {
 	glog.V(5).Infof("GraphBuilder process object: %s/%s, namespace %s, name %s, uid %s, event type %v", event.gvk.GroupVersion().String(), event.gvk.Kind, accessor.GetNamespace(), accessor.GetName(), string(accessor.GetUID()), event.eventType)
 	// Check if the node already exsits
 	existingNode, found := gb.uidToNode.Read(accessor.GetUID())
+	if found {
+		// this marks the node as having been observed via an informer event
+		// 1. this depends on graphChanges only containing add/update events from the actual informer
+		// 2. this allows things tracking virtual nodes' existence to stop polling and rely on informer events
+		existingNode.markObserved()
+	}
 	switch {
 	case (event.eventType == addEvent || event.eventType == updateEvent) && !found:
 		newNode := &node{

--- a/pkg/controller/garbagecollector/graph_builder.go
+++ b/pkg/controller/garbagecollector/graph_builder.go
@@ -382,10 +382,7 @@ func (gb *GraphBuilder) addDependentToOwners(n *node, owners []metav1.OwnerRefer
 		ownerNode, ok := gb.uidToNode.Read(owner.UID)
 		if !ok {
 			// Create a "virtual" node in the graph for the owner if it doesn't
-			// exist in the graph yet. Then enqueue the virtual node into the
-			// attemptToDelete. The garbage processor will enqueue a virtual delete
-			// event to delete it from the graph if API server confirms this
-			// owner doesn't exist.
+			// exist in the graph yet.
 			ownerNode = &node{
 				identity: objectReference{
 					OwnerReference: owner,
@@ -395,9 +392,15 @@ func (gb *GraphBuilder) addDependentToOwners(n *node, owners []metav1.OwnerRefer
 			}
 			glog.V(5).Infof("add virtual node.identity: %s\n\n", ownerNode.identity)
 			gb.uidToNode.Write(ownerNode)
-			gb.attemptToDelete.Add(ownerNode)
 		}
 		ownerNode.addDependent(n)
+		if !ok {
+			// Enqueue the virtual node into attemptToDelete.
+			// The garbage processor will enqueue a virtual delete
+			// event to delete it from the graph if API server confirms this
+			// owner doesn't exist.
+			gb.attemptToDelete.Add(ownerNode)
+		}
 	}
 }
 

